From b557cf5e7d2b0e5335e74c0b4c9eda7d0bf8078b Mon Sep 17 00:00:00 2001
From: Christophe Chapuis <chris.chapuis@gmail.com>
Date: Wed, 12 Aug 2026 22:12:00 +0200
Subject: [PATCH 12/17] WebOSSurfaceItem: close Wayland client fallback on
 Close

 request

When external native apps, like SDL apps, are started on LuneOS, they
don't implement the webOS Wayland extension. This means no shell
surface is created, and we have no means of closing the window.
In this case, just request the Wayland client to close, as a fallback.

Signed-off-by: Christophe Chapuis <chris.chapuis@gmail.com>
Upstream-Status: Pending
---
 modules/weboscompositor/webossurfaceitem.cpp | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/modules/weboscompositor/webossurfaceitem.cpp b/modules/weboscompositor/webossurfaceitem.cpp
index 35a43ee..457f659 100644
--- a/modules/weboscompositor/webossurfaceitem.cpp
+++ b/modules/weboscompositor/webossurfaceitem.cpp
@@ -1105,7 +1105,10 @@ void WebOSSurfaceItem::close()
         sendCloseToGroupItems();
         m_shellSurface->close();
     } else {
-        qWarning() << "No webos shell surface exist, cannot close";
+        qWarning() << "No webos shell surface exist, will close entire client !";
+        if (surface() && surface()->client()) {
+            surface()->client()->close();
+        }
     }
 }
 
